src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -224,59 +235,92 @@ class TezosNodeOperator(val node: TezosRPCInterface)(implicit executionContext:
   }
 
   /**
+    * Given the blockchain network and the hash of the block where the fork was detected,
+    * collect the list of blocks missed during the fork in reverse order.
+    *
+    * Note: We switch from future to IO to utilize the fs2 Stream library effectively,
+    * for concision and efficiency purposes. We may want to consider using IO instead
+    * of Future in general.
     *
-    * @param network
-    * @param hash
+    * @param network Current Blockchain network we're getting blocks from
+    * @param hash Hash of the block where the fork was detected
     * @return Future of List of Blocks that were missed during the Fork, in reverse order
     */
   def followFork(
                   network: String,
-                  hash: String
-                  )(implicit ec: ExecutionContext): Future[List[Block]] = {
-    lazy val offsets: Stream[Int] = Stream.from(1)
-
-    val blockExists: Block => Boolean =
-      block => Await(TezosDatabaseOperations.blockExists(block.metadata.hash))
-    val blockInInvalidatedBlocks: Block => Boolean =
-      block => Await(TezosDatabaseOperations.blockExistsInInvalidatedBlocks(block.metadata.hash))
-
-    val predicate: Future[Block] => (Boolean, Boolean) =
-      futureBlock => {
-        val block = Await.result[Block](futureBlock, Duration.apply(awaitTimeInSeconds, SECONDS))
-        (blockExists(block), blockInInvalidatedBlocks(block))
+                  hash: BlockHash
+                ): Future[List[BlockWithAction]] = {
+
+    //Move some function to utils
+    def futureToIO[A](f: Future[A]): IO[A] = IO.fromFuture(IO(f))
```

Comment:
I think the `f` parameter should better be _by-name_ as in `f: => Future[A]`

This way the future won't start computing until the `IO` result gets evaluated (which is how you make it functionally pure)
Not critical but it might break assumptions down the road